cuda/components/memory.cuh

```diff
@@ -0,0 +1,804 @@
+{
+// see
+// https://github.com/NVIDIA/cutlass/blob/
+//     6fc5008803fe4e81b81a836fcd3a88258f4e5bbf/
+//     include/cutlass/arch/memory_sm75.h#L90
+// for reasoning behind this implementation
+#if (!defined(__clang__) && __CUDACC_VER_MAJOR__ >= 11)
+    return static_cast<uint32>(__cvta_generic_to_shared(ptr));
+#elif (!defined(__clang__) && CUDACC_VER_MAJOR__ == 10 && \
+       __CUDACC_VER_MINOR__ >= 2)
+    return __nvvm_get_smem_pointer(ptr);
+#else
+    uint32 smem_ptr;
+    asm("{{ .reg .u64 smem_ptr; cvta.to.shared.u64 smem_ptr, %1; cvt.u32.u64 "
+        "%0, smem_ptr; }}"
+        : "=r"(smem_ptr)
+        : "l"(ptr));
+    return smem_ptr;
+#endif
+}
+
+
+__device__ __forceinline__ uint32 membar_acq_rel()
+{
+#if __CUDA_ARCH__ < 700
+    asm volatile("membar.gl;" ::: "memory");
+#else
+    asm volatile("fence.acq_rel.gpu;" ::: "memory");
+#endif
+}
+
+
+__device__ __forceinline__ uint32 membar_acq_rel_shared()
+{
+#if __CUDA_ARCH__ < 700
+    asm volatile("membar.cta;" ::: "memory");
+#else
+    asm volatile("fence.acq_rel.cta;" ::: "memory");
+#endif
+}
+
+
+#include "common/cuda_hip/components/memory.hpp.inc"
+
+
+__device__ __forceinline__ int32 load_relaxed_shared(const int32* ptr)
+{
+    int32 result;
+#if __CUDA_ARCH__ < 700
+    asm volatile("ld.volatile.shared.b32 %0, [%1];"
+                 : "=r"(result)
+                 : "r"(convert_generic_ptr_to_smem_ptr(const_cast<int32*>(ptr)))
+                 : "memory");
+#else
+    asm volatile("ld.relaxed.cta.shared.b32 %0, [%1];"
+                 : "=r"(result)
+                 : "r"(convert_generic_ptr_to_smem_ptr(const_cast<int32*>(ptr)))
+                 : "memory");
+#endif
+
+    return result;
+}
+
+
+__device__ __forceinline__ void store_relaxed_shared(int32* ptr, int32 result)
+{
+#if __CUDA_ARCH__ < 700
+    asm volatile("st.volatile.shared.b32 [%0], %1;" ::"r"(
+                     convert_generic_ptr_to_smem_ptr(ptr)),
+                 "r"(result)
+                 : "memory");
+#else
+    asm volatile("st.relaxed.cta.shared.b32 [%0], %1;" ::"r"(
+                     convert_generic_ptr_to_smem_ptr(ptr)),
+                 "r"(result)
+                 : "memory");
+#endif
+}
+
+
+__device__ __forceinline__ int64 load_relaxed_shared(const int64* ptr)
+{
+    int64 result;
+#if __CUDA_ARCH__ < 700
+    asm volatile("ld.volatile.shared.b64 %0, [%1];"
+                 : "=l"(result)
+                 : "r"(convert_generic_ptr_to_smem_ptr(const_cast<int64*>(ptr)))
+                 : "memory");
+#else
+    asm volatile("ld.relaxed.cta.shared.b64 %0, [%1];"
+                 : "=l"(result)
+                 : "r"(convert_generic_ptr_to_smem_ptr(const_cast<int64*>(ptr)))
+                 : "memory");
+#endif
+
+    return result;
+}
+
+
+__device__ __forceinline__ void store_relaxed_shared(int64* ptr, int64 result)
+{
+#if __CUDA_ARCH__ < 700
+    asm volatile("st.volatile.shared.b64 [%0], %1;" ::"r"(
```

Comment:
```suggestion
    asm volatile("st.volatile.shared.s64 [%0], %1;" ::"r"(
```

Reply:
good catch, though it probably doesn't make a difference. this is what `libcudacxx` uses in their atomic loads/stores, they only use signed/unsigned types for arithmetic operations in atomics (+,-,max,min).